include/cuco/probing_scheme.cuh

```diff
@@ -175,6 +177,159 @@ class double_hashing : private detail::probing_scheme_base<CGSize> {
+  /**
+   *@brief Constructs perfect probing scheme with the hasher callable.
+   *
+   * @param hash Hasher
+   */
+  __host__ __device__ constexpr perfect_probing(Hash const& hash = {});
+
+  /**
+   *@brief Makes a copy of the current probing method with the given hasher
+   *
+   * @tparam NewHash New hasher type
+   *
+   * @param hash Hasher
+   *
+   * @return Copy of the current probing method
+   */
+  template <typename NewHash>
+  [[nodiscard]] __host__ __device__ constexpr auto with_hash_function(
+    NewHash const& hash) const noexcept;
+
+  /**
+   * @brief Operator to return a probing iterator
+   *
+   * @tparam ProbeKey Type of probing key
+   * @tparam Extent Type of extent
+   *
+   * @param probe_key The probing key
+   * @param upper_bound Upper bound of the iteration
+   * @return An iterator whose value_type is convertible to slot index type
+   */
+  template <typename ProbeKey, typename Extent>
+  __host__ __device__ constexpr auto operator()(ProbeKey const& probe_key,
+                                                Extent upper_bound) const noexcept;
+
+  /**
+   * @brief Operator to return a CG-based probing iterator
+   *
+   * @tparam ProbeKey Type of probing key
+   * @tparam Extent Type of extent
+   *
+   * @param g the Cooperative Group to generate probing iterator
+   * @param probe_key The probing key
+   * @param upper_bound Upper bound of the iteration
+   * @return An iterator whose value_type is convertible to slot index type
+   */
+  template <typename ProbeKey, typename Extent>
+  __host__ __device__ constexpr auto operator()(
+    cooperative_groups::thread_block_tile<cg_size> const& g,
+    ProbeKey const& probe_key,
+    Extent upper_bound) const noexcept;
+
+  /**
+   * @brief Perfect probing iterator class.
+   *
+   * @tparam Extent Type of Extent
+   */
+  template <typename Extent>
+  class probing_iterator {
+   public:
+    using extent_type = Extent;                            ///< Extent type
+    using size_type   = typename extent_type::value_type;  ///< Size type
+
+    /**
+     * @brief Constructs an probing iterator
+     *
+     * @param start Iteration starting point
+     * @param upper_bound Upper bound of the iteration
+     */
+    __host__ __device__ constexpr probing_iterator(size_type start,
+                                                   extent_type upper_bound) noexcept
+      : curr_index_{start}, upper_bound_{upper_bound}
+    {
+    }
+
+    /**
+     * @brief Dereference operator
+     *
+     * @return Current slot index
+     */
+    __host__ __device__ constexpr auto operator*() const noexcept { return curr_index_; }
+
+    /**
+     * @brief Prefix increment operator
+     *
+     * @return Current iterator
+     */
+    __host__ __device__ constexpr auto operator++() noexcept
+    {
+      // TODO: This is still getting called, and should never really be used
+      curr_index_ = (curr_index_ + 1) % upper_bound_;
+      // curr_index_ = 0;
+      return *this;
```

Comment:
```suggestion
  static_assert(false, "perfect probing doesn't increment");
```
question: do we need to probe (increment the pointer) in the case of perfect hashing?

Reply:
Never.  Good catch @PointKernel !

If this function is getting called called, there is a programming error at the call site?  Hmm, except for multisets and multihashs...

I could see how this function might get called if the cooperative group size is not 1.  perfect hashing should disallow cooperative groups, too.

I like the `static_assert` here instead of deleting the function.  It provides more information to the programmer.  I would like more detail, either in the blurb or in a comment, about why perfect probing doesn't even need incrementing.

```cuda
static_assert(false, "perfect probing doesn't support incrementing because incrementing is only necessary when a slot is found to be full but not matching, which is impossible");
```

Go through the code and figure out why this is getting called.  If it's due to the cooperative groups stuff then those test cases must be deleted.  perfect probing is not compatible with CG.  If it's due to multiset or multihash then delete those test cases, too.  Not compatible.

For all other cases, there is a bug in the code and it should be fixed.